### PR TITLE
Remove dead code after destination propagation

### LINKED ISSUE #105428: Broken MIR in encoding_rs related to MIR inlining + DestProp
This was originally https://github.com/rust-lang/rust/issues/105344#issuecomment-1340153677

---

Trying to compile https://github.com/hsivonen/encoding_rs `5931e21bc20bad0335964f441da0c4309401107a` with these flags:
```
$ RUSTFLAGS="-Zmir-opt-level=3 -Zvalidate-mir -Zinline-mir=yes" cargo +nightly test
```

Produces this:
```
error: internal compiler error: broken MIR in Item(WithOptConstParam { did: DefId(0:1315 ~ encoding_rs[5e1f]::utf_16::{impl#0}::decode_to_utf16_raw), const_param_did: None }) (before pass SimplifyCfg-final) at bb218[0]:
                                encountered overlapping memory in `Call` terminator: _454 = bswap::<u16>(move _454) -> bb163
  --> src/utf_16.rs:85:21
   |
85 |                     dest.copy_utf16_from::<LittleEndian>(&mut source)
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: delayed at compiler/rustc_const_eval/src/transform/validate.rs:784:26
```

Removing `-Zinline-mir=yes` or `-Zmir-opt-level=3` removes the ICE.

---

searched nightlies: from nightly-2022-01-01 to nightly-2022-12-05
regressed nightly: nightly-2022-11-28
searched commit range: https://github.com/rust-lang/rust/compare/80a96467ec5675e9f69683b5c075a8b15950c341...1eb62b1235fd77200e6bd967d70e83c0f2497233
regressed commit: https://github.com/rust-lang/rust/commit/0e9eee68111ec742c3d60bb96078994c494a59fb

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.5</summary>

Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --script script --start 2022-01-01 --end 2022-12-05 
```
</details>

### PR DESCRIPTION
**Pull request by tmiasko**
Fixes #105428.

cc @JakobDegen

### CODE REVIEW

**Comment by JakobDegen** [comment]
Ah, yeah, that explains it. Good catch!

**Comment by oli-obk** [comment]
r? @oli-obk 

I don't see a minimal repro on the issue. Do you know how to trigger it and can add a test?

**Comment by tmiasko** [comment]
Custom MIR doesn't support function calls yet, so I don't think I can add a test (reproducing the issue requires dead code of specific form to exist precisely at the time when destination propagation runs).

**Comment by JakobDegen** [comment]
> and assert there are none for DestProp to be sound?

Bit of a nitpick, but probably worth pointing out: This is not strictly speaking a soundness bug. Dest prop is introducing UB in dead code, which is "fine."

> Custom MIR doesn't support function calls yet, so I don't think I can add a test (reproducing the issue requires dead code of specific form to exist precisely at the time when destination propagation runs).

I can add support for this this weekend. We can delay this PR until then, or I can just add a test in a follow up at that time.

File: compiler/rustc_mir_transform/src/dest_prop.rs [hunk 1]
```diff
@@ -238,6 +239,12 @@ impl<'tcx> MirPass<'tcx> for DestinationPropagation {
             apply_merges(body, tcx, &merges, &merged_locals);
         }
 
+        if round_count != 0 {
+            // Merging can introduce ill-formed overlap between moved arguments and/or call
+            // destination in an unreachable code.
+            remove_dead_blocks(tcx, body);
```
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/dest_prop.rs:245]: What is the structural reason that this can only happen in dead code?

Comment (JakobDegen) [reply compiler/rustc_mir_transform/src/dest_prop.rs:245]: The logic that prevents us from introducing this kind of UB is [here](https://github.com/rust-lang/rust/blob/7632db0e87d8adccc9a83a47795c9411b1455855/compiler/rustc_mir_transform/src/dest_prop.rs#L454). That loop only iterates over reachable blocks, which means we can accidentally introduce these kinds of ICEs in unreachable blocks.

**Comment by saethlin** [comment]
> Dest prop is introducing UB in dead code, which is "fine."

Can/should we adjust MIR validation to ignore dead code? (if this PR is blocked on testing via custom MIR, such a PR would be too, but still...)

**Comment by JakobDegen** [comment]
> Can/should we adjust MIR validation to ignore dead code?

Not in general. "Well-formed MIR" is something that ought to have a precise definition, and (ignoring that that doesn't exist today) we should not intentionally diverge from that in the validator.

That being said, this particular check should probably be turned off in dead code because the MIR is technically well-formed. Will file an issue

**Comment by tmiasko** [comment]
* Added a regression test for the original issue.

**Comment by JakobDegen** [comment]
LGTM

**Pull request merged**